conda/env/installers/conda.py

```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
     subdirs = IndexedSet(basename(url) for url in _channel_priority_map)
 
     solver_backend = context.plugin_manager.get_cached_solver_backend()
+    assert solver_backend is not None  # get_cached_solver_backend never returns None
     solver = solver_backend(prefix, channels, subdirs, specs_to_add=specs)
     return solver
 
 
-def dry_run(specs, args, env, *_, **kwargs):
-    """Do a dry run of the environment solve"""
+def dry_run(
+    specs: list, args: Namespace, env: Environment, *_, **kwargs
+) -> Environment:
+    """Do a dry run of the environment solve.
+
+    :param specs: Package specifications to install
+    :param args: Command-line arguments
+    :param env: Environment object
+    :return: Solved environment object
+    :rtype: Environment
+    """
     solver = _solve(tempfile.mkdtemp(), specs, args, env, *_, **kwargs)
     pkgs = solver.solve_final_state()
-    solved_env = Environment(
+    return Environment(
         name=env.name, dependencies=[str(p) for p in pkgs], channels=env.channels
     )
-    return solved_env
 
 
-def install(prefix, specs, args, env, *_, **kwargs):
-    """Install packages into an environment"""
+def install(
+    prefix: str, specs: list, args: Namespace, env: Environment, *_, **kwargs
+) -> dict | None:
+    """Install packages into a conda environment.
+
+    This function handles two main paths:
+    1. For environments with explicit_specs (from @EXPLICIT files): Bypasses the solver
+       and directly installs packages using conda.misc.explicit() as required by CEP-23.
+    2. For regular Environment instances: Uses the solver to determine the optimal
+       package set before installation.
+
+    :param prefix: The target installation path for the environment
+    :type prefix: str
```

Comment:
I don't think we need to double annotate the types here if we have type hints already.